tensorflow/core/kernels/diag_op.cc

```diff
@@ -129,24 +130,38 @@ namespace functor {
 template <typename T>
 struct DiagFunctor<CPUDevice, T> {
   EIGEN_ALWAYS_INLINE Status
-  operator() (const CPUDevice& device, const int64 size,
-                   const T* in, T* out) {
-    std::fill(out, out + size * size, T());
-    for (int64 index = 0; index < size; index++) {
-      out[(1 + size) * index] = in[index];
-    }
+  operator() (OpKernelContext* context, const int64 size,
+              const T* in, T* out) {
+    auto subDiag = [in, out, size](int64 start, int64 limit) {
+      std::fill(out + size * start, out + size * limit, T());
+      for (int64 index = start; index < limit; ++index) {
+        out[(1 + size) * index] = in[index];
+      }
+    };
+
+    // Here, 5 is a empirical factor of cost_per_unit.
+    auto worker_threads = *(context->device()->tensorflow_cpu_worker_threads());
+    Shard(worker_threads.num_threads, worker_threads.workers, size,
+        5 * size, subDiag);
     return Status::OK();
   }
 };
 
 template <typename T>
 struct DiagPartFunctor<CPUDevice, T> {
   EIGEN_ALWAYS_INLINE Status
-  operator() (const CPUDevice& device, const int64 size,
-                   const T* in, T* out) {
-    for (int64 index = 0; index < size; index++) {
-      out[index] = in[(1 + size) * index];
-    }
+  operator() (OpKernelContext* context, const int64 size,
+              const T* in, T* out) {
+    auto subDiagPart = [in, out, size](int64 start, int64 limit) {
```

Comment:
perhaps add one line comment here too (although in this case it's more obvious there isn't a different way of sharding).